src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,150 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2016 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import re
+import sys
+import string
+from itertools import chain
+
+import hypothesis.strategies as strats
+from hypothesis.searchstrategy import SearchStrategy
+from hypothesis.internal.compat import hunichr, hrange
+
+
+def strategy_concat(strategies):
+    """given a list of strategies yielding strings return a strategy yielding
+    their concatenation."""
+    return strats.tuples(*strategies).map(lambda x: u"".join(x))
+
+
+class RegexStrategy(SearchStrategy):
+    """Strategy for generating strings matching a regular expression.
+
+    Currently does not support \B, positive and negative lookbehind
+    assertions and conditional matching.
+
+    """
+
+    def __init__(self, pattern):
+        parsed = re.sre_parse.parse(pattern)
+        self.cache = {}
```

Comment:
This cache is for resolving group references, i.e. \1 or (?P<name>...) see l. 115 and l. 126.

Cacheing would indeed only make sense only on a global level. But I don't think I understand typical workloads well enough to have an opinion on its effectivity.